katdal/caldata.py

```diff
@@ -0,0 +1,383 @@
+    -------
+    dict
+        Keys are 'h' and 'v' and values are 0 and 1, in some order
+    """
+    cal_pol_ordering = _get_cal_attr('cal_pol_ordering', katdal_obj)
+    try:
+        cal_pol_ordering = np.array(cal_pol_ordering)
+    except (NameError, SyntaxError):
+        raise
+    except Exception as e:
+        raise CalibrationReadError(str(e))
+    if cal_pol_ordering.shape != (4, 2):
+        raise CalibrationReadError('cal_pol_ordering does not have expected shape')
+    if cal_pol_ordering[0, 0] != cal_pol_ordering[0, 1]:
+        raise CalibrationReadError('cal_pol_ordering[0] is not consistent')
+    if cal_pol_ordering[1, 0] != cal_pol_ordering[1, 1]:
+        raise CalibrationReadError('cal_pol_ordering[1] is not consistent')
+    order = [cal_pol_ordering[0, 0], cal_pol_ordering[1, 0]]
+    if set(order) != set('vh'):
+        raise CalibrationReadError('cal_pol_ordering does not contain h and v')
+    return {order[0]: 0, order[1]: 1}
+
+
+def _get_cal_product(key, katdal_obj, **kwargs):
+    """Loads calibration solutions from a katdal file.
+
+    If an error occurs while loading the data, a warning is printed and the
+    return value is ``None``. Any keyword args are passed to
+    :func:`scipy.interpolate.interp1d` or `ComplexInterpolate1D`.
+
+    Solutions that contain non-finite values are discarded.
+
+    Parameters
+    ----------
+    key : str
+        Name of the telescope state sensor
+
+    Returns
+    -------
+    interp : callable
+        Interpolation function which accepts timestamps and returns
+        interpolated data with shape (time, channel, pol, antenna). If the
+        solution is channel-independent, that axis will be present with
+        size 1.
+    """
+    timestamps, values = _get_cal_attr(key, katdal_obj, sensor=False)
+    values = np.asarray(values)
+
+    if (timestamps[-1] < katdal_obj._timestamps[0]):
+        print('All %s calibration solution ahead of observation, no overlap' % key)
+    elif (timestamps[0] > katdal_obj._timestamps[-1]):
+        print('All %s calibration solution after observation, no overlap' % key)
+
+    if values.ndim == 3:
+        # Insert a channel axis
+        values = values[:, np.newaxis, ...]
+
+    if values.ndim != 4:
+        raise ValueError('Calibration solutions has wrong number of dimensions')
+
+    # only use solutions with valid values, assuming matrix dimensions
+    # (ts, chan, pol, ant)
+    # - all values per antenna must be valid
+    # - all values per polarisation must be valid
+    # - some channels must be valid (you will interpolate over them later)
+    ts_mask = np.isfinite(values).all(axis=-1).all(axis=-1).any(axis=-1)
+    if (ts_mask.sum() / float(len(timestamps))) < 0.7:
+        raise ValueError('no finite solutions')
+    values = values[ts_mask, ...]
+    timestamps = timestamps[ts_mask]
+
+    if values.shape[1] > 1:
+        # Only use channels selected in h5 file
+        values = values[:, katdal_obj.channels, ...]
```

Comment:
I would rather not enforce this to be over all channels, if you have a significant number of channels, this calculation can be very expensive.
my preference is to only do what you need.
channel selections uses slices and in general are regular, the interpolation function also evaluates the nan-ness of the values before taking action

Reply:
I think the solution is to separate the set of channels used to define the interpolation (which needs to be the full range) from the set on which you sample i.e. the channel selection should be passed down into interpolate_nans_1d.

Reply:
Had a good look at the code, at the moment it will use what is desired under the __keep flags. That is the lazy way. This is going to be very involved to change.

I feel this will be a bad implementation, but will have a more elegant solution outside the lazy implementation. So if you do not mind, I will keep this note for the redesign and I also feel that the calibration and other more specific "higher level" science processing should be moved to a package that extends katdal.